File: DGEMM/dgemm.jl
using Base.Threads
using LinearAlgebra
using BenchmarkTools

function naive_dgemm!(C, A, B)
    N = size(A, 1)
    @threads for i in 1:N
        for j in 1:N
            acc = 0.0
            for k in 1:N
                @inbounds acc += A[i, k] * B[k, j]
            end
            @inbounds C[i, j] = acc
        end
    end
    return C
end

function optimized_dgemm!(C, A, B, block_size)
    N = size(A, 1)
    Bt = transpose(B)
    fill!(C, 0.0)
    
    @threads for ii in 1:block_size:N
        i_end = min(ii + block_size - 1, N)
        for jj in 1:block_size:N
            j_end = min(jj + block_size - 1, N)
            for kk in 1:block_size:N
                k_end = min(kk + block_size - 1, N)
                for i in ii:i_end
                    for j in jj:j_end
                        acc = 0.0
                        @inbounds @simd for k in kk:k_end
                            acc += A[i, k] * Bt[j, k]
                        end
                        @inbounds C[i, j] += acc
                    end
                end
            end
        end
    end
    return C
end

function validate(C1, C2; atol=1e-6)
    return isapprox(C1, C2; atol=atol)
end

function benchmark()
    sizes = [512, 1024, 2048]
    blocks = [32, 64, 128]
    println("| Size | Block | Method    | Time (s) | GFLOPS  | Valid |")
    println("|------|-------|-----------|----------|---------|-------|")
    
    for n in sizes
        println(n)
        A = rand(n, n)
        B = rand(n, n)
        C_ref = A * B
        C_naive = similar(A)
        C_opt = similar(A)
        
        # Benchmark BLAS
        blas_time = @belapsed $A * $B
        blas_gflops = 2n^3 / blas_time / 1e9
        println("| $n | - | BLAS | $(round(blas_time, digits=4)) | $(round(blas_gflops, digits=2)) | true |")
        
        # Benchmark optimized versions
        for bs in blocks
            opt_time = @belapsed optimized_dgemm!($C_opt, $A, $B, $bs)
            opt_gflops = 2n^3 / opt_time / 1e9
            valid = validate(C_opt, C_ref)
            println("| $n | $bs | Optimized | $(round(opt_time, digits=4)) | $(round(opt_gflops, digits=2)) | $valid |")
            fill!(C_opt, 0.0)
        end
        println(n)
        # Benchmark naive implementation
        if n <= 1024
            naive_time = @belapsed naive_dgemm!($C_naive, $A, $B)
            naive_gflops = 2n^3 / naive_time / 1e9
            valid = validate(C_naive, C_ref)
            println("| $n | - | Naive | $(round(naive_time, digits=4)) | $(round(naive_gflops, digits=2)) | $valid |")
        end
    end
end

# Run benchmark comparison
benchmark()

File: DGEMM/dgemm.py
import numpy as np
import numba as nb
import time
import os
from concurrent.futures import ThreadPoolExecutor

# Set environment variables for thread control
os.environ['OMP_NUM_THREADS'] = '4'
os.environ['MKL_NUM_THREADS'] = '1'

@nb.njit(nogil=True, parallel=True, fastmath=True)
def dgemm_parallel(A, B, C, alpha=1.0, beta=0.0, block_size=64):
    n = A.shape[0]
    Bt = np.empty_like(B)
    
    # Transpose B for better cache utilization
    for i in nb.prange(n):
        for j in range(n):
            Bt[j, i] = B[i, j]
    
    # Blocked matrix multiplication with parallelization
    for ii in nb.prange(0, n, block_size):
        for jj in range(0, n, block_size):
            for kk in range(0, n, block_size):
                i_end = min(ii + block_size, n)
                j_end = min(jj + block_size, n)
                k_end = min(kk + block_size, n)
                
                for i in range(ii, i_end):
                    for j in range(jj, j_end):
                        tmp = 0.0
                        for k in range(kk, k_end):
                            tmp += A[i, k] * Bt[j, k]
                        C[i, j] = alpha * tmp + beta * C[i, j]

@nb.njit(nogil=True, parallel=True)
def dgemm_naive(A, B, C, alpha=1.0, beta=0.0):
    n = A.shape[0]
    for i in nb.prange(n):
        for j in range(n):
            tmp = 0.0
            for k in range(n):
                tmp += A[i, k] * B[k, j]
            C[i, j] = alpha * tmp + beta * C[i, j]

def validate(ref, test, tol=1e-6):
    return np.allclose(ref, test, atol=tol)

def benchmark(matrix_sizes=[512, 1024, 2048], block_sizes=[32, 64, 128]):
    results = []
    
    for size in matrix_sizes:
        A = np.random.rand(size, size)
        B = np.random.rand(size, size)
        C_numpy = np.zeros((size, size))
        C_parallel = np.zeros((size, size))
        C_naive = np.zeros((size, size))
        
        # Numpy reference implementation
        start = time.perf_counter()
        ref = A @ B
        numpy_time = time.perf_counter() - start
        numpy_gflops = (2 * size**3) / numpy_time / 1e9
        
        # Warm-up Numba JIT
        dgemm_parallel(A, B, C_parallel)
        dgemm_naive(A, B, C_naive)
        
        # Benchmark parallel implementation with different block sizes
        for bs in block_sizes:
            C_parallel.fill(0.0)
            start = time.perf_counter()
            dgemm_parallel(A, B, C_parallel, bs)
            parallel_time = time.perf_counter() - start
            parallel_gflops = (2 * size**3) / parallel_time / 1e9
            valid = validate(ref, C_parallel)
            
            results.append((
                size, bs, 'Parallel',
                parallel_time, parallel_gflops, valid
            ))
        
        # Benchmark naive implementation
        C_naive.fill(0.0)
        start = time.perf_counter()
        dgemm_naive(A, B, C_naive)
        naive_time = time.perf_counter() - start
        naive_gflops = (2 * size**3) / naive_time / 1e9
        valid = validate(ref, C_naive)
        
        results.append((
            size, 'N/A', 'Naive',
            naive_time, naive_gflops, valid
        ))
        
        # Numpy results
        results.append((
            size, 'N/A', 'NumPy',
            numpy_time, numpy_gflops, True
        ))
    
    # Print results
    print(f"{'Size':<8} {'Block':<6} {'Method':<8} {'Time (s)':<10} {'GFLOPS':<10} {'Valid'}")
    print("="*60)
    for res in results:
        print(f"{res[0]:<8} {str(res[1]):<6} {res[2]:<8} {res[3]:<10.4f} {res[4]:<10.2f} {res[5]}")

if __name__ == "__main__":
    # Benchmark different configurations
    benchmark(
        matrix_sizes=[512, 1024, 2048],
        block_sizes=[32, 64, 128]
    )
